src/hotspot/share/runtime/mutex.cpp

```diff
@@ -137,14 +137,14 @@ void Mutex::lock_without_safepoint_check() {
 
 
 // Returns true if thread succeeds in grabbing the lock, otherwise false.
-
+// Checking for a NULL owner avoids the call into platform code and hides the
+// potential difference in recursive locking behaviour on some platforms.
 bool Mutex::try_lock() {
   Thread * const self = Thread::current();
   // Some safepoint_check_always locks use try_lock, so cannot check
   // safepoint state, but can check blocking state.
   check_block_state(self);
-  if (_lock.try_lock()) {
-    assert_owner(NULL);
+  if (_owner == NULL && _lock.try_lock()) {
```

Comment:
`_owner != self` may be a bit clearer, at the expense of maybe unnecessarily calling into platform code (but before, we did that unconditionally).